### PR TITLE
Use fastOutSlowIn curve for transitions on iOS when not controlled by…

### PR DESCRIPTION
**Pull request by mpcomplete**
… user gesture.

BUG=https://github.com/flutter/flutter/issues/5599

### CODE REVIEW

File: packages/flutter/lib/src/material/page.dart [hunk 1]
```diff
@@ -48,13 +48,14 @@ class _CupertinoPageTransition extends AnimatedWidget {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/page.dart]: while we're here, can you rename _kTween to something that says what it is (e.g. _kHorizontalSlideTween) and make it a constant? (can it be a constant?)

File: packages/flutter/lib/src/material/page.dart [hunk 2]
```diff
@@ -88,12 +89,25 @@ class AnimationMean extends CompoundAnimation<double> {
...
+
+  Curve curve;
 
   @override
   double transform(double t) {
-    if (t > 0.5)
-      return (t - 0.5) / 3.0 + 0.5;
+    // The range is divided in 2 halves: [0.0, 0.5] applies to the top-most
+    // route, while [0.5, 1.0] applies to the one underneath it. We scale both
+    // ranges to [0.0, 1.0] when applying transformations.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/page.dart]: i don't understand this comment at all :-(

File: packages/flutter/lib/src/material/page.dart [hunk 3]
```diff
@@ -233,6 +248,10 @@ class MaterialPageRoute<T> extends PageRoute<T> {
         );
       case TargetPlatform.iOS:
         return new _CupertinoPageTransition(
+          // Use a linear curve when controlled by a user gesture. This ensures
+          // the animation tracks the user's finger 1:1.
+          // See https://github.com/flutter/flutter/issues/5664
+          curve: userGesture ? null : Curves.fastOutSlowIn,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/page.dart]: on iOS what happens when you fling it? Does it then follow a curve or simulation?

what happens when you let go? Won't this make the page jump?

**Comment by Hixie** [comment]
LGTM but I don't fully understand it.

File: packages/flutter/lib/src/material/page.dart (hunk 3 above)
Comment (mpcomplete) [inline packages/flutter/lib/src/material/page.dart]: I didn't implement fling. I'll have to add support for that. (It just treats fling same as stop-and-release.)

userGesture is true if the transition was started with a user gesture. It remains true throughout the transition, even after release.

File: packages/flutter/lib/src/material/page.dart (hunk 2 above)
Comment (mpcomplete) [inline packages/flutter/lib/src/material/page.dart]: Is this any clearer:

    // The input [t] is the average of the previous and current route's
    // animations. So, when [t] is in the range [0.0, 0.5], we're dealing with
    // the top-most route. Likewise, [0.5, 1.0] applies to the route underneath
    // it (to the left of the screen).
    // To make the curves apply properly, we scale [t] to the range [0.0, 1.0]
    // before applying transformations.

File: packages/flutter/lib/src/material/page.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/page.dart]: Ah, ok. We'll probably have to curve that case to in due course.

File: packages/flutter/lib/src/material/page.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/page.dart]: I don't understand why this is true. Suppose you popped a route, then 20ms later popped another. Both routes would be moving at the same time. There's no guarantee that only one of the animations will be changing at any particular time.

Comment (mpcomplete) [inline packages/flutter/lib/src/material/page.dart]: Ah, so maybe it's not true :). Probably why it didn't make sense.

It's hard for me to reason about what should or would happen when there are multiple transitions happening at once, so I only focused on the simple case of a push or pop. My goal with this code was to have the topmost page have fast transition from all-the-way-to-the-right to onscreen, while the page underneath has a slower transition from onscreen to about-a-third-to-the-left. Probably clearer if you watch what the iOS transition is doing. But in the simple case, t=0.5 represents the case when the topmost page is fully on screen, and the transition is settled.

**Pull request merged**